### ISSUE TITLE
Possible memory leak when building training files

### ISSUE THREAD

**Issue by H-Bluhm** [tesseract-ocr/tesseract#1999]
### Environment

* **Tesseract Version**: 4.0-rc3, 4.0-rc2, 4.0-rc1
* **Platform**: Linux 9000119697 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

Cut training text into smaller chunks (~20000 lines each, ~1.3mb) to build training files on smaller computers in parallel.
Tried building frk training files using:
```
src/training/tesstrain.sh \
--lang frk \
--linedata_only \
--noextract_font_properties \
--fonts_dir ~/frk_fonts/ \
--langdata_dir ~/langdata_lstm/ \
--tessdata_dir ~/tessdata/ \
--output_dir ~/tessOutput/
```
### Current Behavior:
Memory usage increases rapidly (total >15gb when process was killed by kernel)
### Expected Behavior:
With same training text until 4.0-beta.4 memory usage increased at a much lower rate (max usage <4gb)

**Comment by Shreeshrii** [comment]
I am also noticing similar behavior when running training for Devanagari using a subset of training_text from langdata_lstm. The process is killed with signal 9.

**Comment by eighttails** [comment]
I'm also just about to report the same issue and have funished bisecting.
Maybe related to this commit.
https://github.com/tesseract-ocr/tesseract/commit/345e5ee

**Comment by eighttails** [comment]
tesstrain.sh continues even if text2image crashed.
After that lstmtraining generates error message because text2image died without generating box files.

**Comment by stweil** [comment]
This looks like a bug and a regression if beta.4 still was fine. So we have to decide whether fixing it is required for 4.0.0.

**Comment by amitdo** [comment]
It depends on how quickly you can find the faulty commit...

**Comment by Shreeshrii** [comment]
A number of people use training, specially for finetuning of non-Latin languages which have been trained at Google with fewer fonts.

While there are many issues with training that are on hold right now, I think this regression should be fixed before 4.0.0.

**Comment by Shreeshrii** [comment]
@eighttails has identified a commit in https://github.com/tesseract-ocr/tesseract/issues/1999#issuecomment-431025048

**Comment by amitdo** [comment]
Another training issue:

https://github.com/tesseract-ocr/tesseract/issues/1052

https://github.com/tesseract-ocr/tesseract/issues/1700#issuecomment-399766478

**Comment by zdenop** [comment]
@eighttails : you are right. Problem is that if PangoFontMap is created with pango_cairo_font_map_get_default() it [must not be freed](https://developer.gnome.org/pango/stable/pango-Cairo-Rendering.html#pango-cairo-font-map-get-default) , but if it is created with pango_cairo_font_map_new_for_font_type, [it should be freed](https://developer.gnome.org/pango/stable/pango-Cairo-Rendering.html#pango-cairo-font-map-new-for-font-type)...

**Comment by zdenop** [comment]
Please check

**Comment by mgeerdsen** [comment]
Memory resumption is looking good again (text2image 4.0.0-rc3-41-g0a42c0).

However I do not get any text rendered into the images (just blank pages), whereas this worked in the same setup with 4.0.0-beta.4. I will look into this and open up a separate issue for it.
